app/common/util/src/main/java/io/syndesis/common/util/cache/LRUSoftCache.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.syndesis.common.util.cache;
+
+import java.lang.ref.Reference;
+import java.lang.ref.SoftReference;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A special cache using soft references to hold data, to allow the garbage collector to reclaim space if the application
+ * is suffering for high memory usage.
+ */
+public class LRUSoftCache<K, V> implements Cache<K, V> {
+
+    private Cache<K, Reference<V>> cache;
+
+    public LRUSoftCache(int maxElements) {
+        this.cache = new LRUDefaultCache<>(maxElements);
+    }
+
+    @Override
+    public V get(K key) {
+        return getOptional(key)
+            .orElse(null);
+    }
+
+    @Override
+    public Optional<V> getOptional(K key) {
+        return cache.getOptional(key)
+            .flatMap(ref -> Optional.ofNullable(ref.get()));
+    }
+
+    @Override
+    public Set<K> keySet() {
+        return cache.keySet()
+            .stream()
+            .filter(k -> cache.getOptional(k).flatMap(ref -> Optional.ofNullable(ref.get())).isPresent())
+            .collect(Collectors.toSet());
+    }
+
+    @Override
+    public Collection<V> values() {
+        return cache.values()
+            .stream()
+            .map(ref -> Optional.ofNullable(ref.get()))
```

Comment:
can't this be just cahe.values().stream().filter(ref -> ref.get() != null) ?

Reply:
Close.. "getting optional value then filtering" is not equal to "filtering then getting value". In the latter case, the value may disappear..

Reply:
it can be map(ref::get).filter(Objects::notNull)

Reply:
Right, I always fear passing nulls to `map` functions, but I think it only causes troubles in `Optional`, not in `Stream`s. Btw, I'm adding a Guava impl of the Cache. I don't see problems in using it but in any case we can switch back to this one in case of issues.